Reducing redundant state changes

js6i opened the pull request:
Here's a couple of patches that in my tests can reduce Metal command buffer encoding time pretty significantly in some cases, by avoiding encoding things redundantly.
There's another one (not included) to allocate MVKCommands into a linear buffer, which speeds things up a little and reduces freeing them to a trivial "count = 0;" (which otherwise takes some time of the order of 1ms from what I've seen), but there's the complication of a handful commands which actually do things in their destructor, which I haven't resolved yet.

cdavis5e reviewed (changes requested):
I was wondering about this myself. "Why are we emitting all these redundant state changes?" Hopefully this will help with the performance problems other people have been seeing.

Just one small problem... (Besides the fact that #1595 changed the same part of the code.)

cdavis5e commented:
@ZoeyCluff Does this change help with the performance problems you've been seeing? (You may have trouble applying it because of #1595.)

billhollings commented on MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.h:
```diff
@@ -387,11 +388,12 @@ class MVKResourcesCommandEncoderState : public MVKCommandEncoderState {
+                    *iter = db;
                 return;
             }
         }
```
Per the comment below by @cdavis5e, regarding side effects in `areEqual()`, I'm wondering if it might be a  good idea to merge this logic into something like an `updateIfReplaced(b)` function on the binding structs, and call `areEqual()` and `markDirty()`, plus handle the additional specialized buffer logic for `justOffset` within that `updateIfReplaced(b)` function. Then we'd end up with something here like:

```
for (auto iter = bindings.begin(), end = bindings.end(); iter != end; ++iter) {
    if (iter->updateIfReplaced(b))
        return;
}
```

with no need for the `db` copy, because the `updateIfReplaced()` function in `*iter` would call it's own `markDirty()`. And I've assumed `areEqual() could presumably include the `index == b.index` check too.

billhollings commented on MoltenVK/MoltenVK/Commands/MVKMTLResourceBindings.h:
```diff
@@ -34,33 +34,64 @@ typedef struct MVKMTLTextureBinding {
+        if (a->offset == b->offset)
+            return true;
+        if (!a->isInline && !b->isDirty)
+            a->justOffset = true;
```
Agreed. I've added some thoughts above about refactoring this part (plus other logic) into a possible `updateIfReplaced()` function.

billhollings reviewed (changes requested):
Thanks so much for submitting this! It is definitely a helpful performance improvement!

Need changes to align with #1595, and I've included a suggestion for refactoring, adding to @cdavis5e 's thoughts.

js6i commented:
Thanks, I changed the isEqual() procedure to update(), which makes more sense. Also changed the justOffset logic a little bit - setting the flag doesn't have to be predicated on the binding not being inline, as that is checked during encoding. Additionally, if the old binding already was justOffset dirty, we can preserve that (pretty much an edge case, but hey).

FleetAdmiralButter commented:
Hi @js6i,

Apologies for butting in - I gave this PR a spin with Final Fantasy XIV but unfortunately, it completely breaks lighting/shadows in the game:

https://www.youtube.com/watch?v=WTcn3lbWeOY
https://www.youtube.com/watch?v=bpM36Q_CMDk

I'll try this out with another DXVK game shortly and see if it exhibits the same issue.

js6i commented:
> Hi @js6i,
> 
> Apologies for butting in - I gave this PR a spin with Final Fantasy XIV but unfortunately, it completely breaks lighting/shadows in the game:
> 
> https://www.youtube.com/watch?v=WTcn3lbWeOY https://www.youtube.com/watch?v=bpM36Q_CMDk
> 
> I'll try this out with another DXVK game shortly and see if it exhibits the same issue.

Interesting, I tested it with FFXIV, but using wined3d and not DXVK. Note that I did a second force push after the first one, where I got the update backwards, but I don't see how that could make things worse. I'll try with DXVK. Thanks!

FleetAdmiralButter commented:
> Interesting, I tested it with FFXIV, but using wined3d and not DXVK. Note that I did a second force push after the first one, where I got the update backwards, but I don't see how that could make things worse. I'll try with DXVK. Thanks!

Thanks for looking into this @js6i!

Just tested with a couple NieR games I have in CrossOver and they don't seem to exhibit the same lighting issue fortunately:

![image](https://user-images.githubusercontent.com/37036084/170044189-f9296b22-8ba0-46fd-bf7a-f255629a06cf.png)
![image](https://user-images.githubusercontent.com/37036084/170044251-f48c42af-a344-4062-bbfa-fc9c55c68cdd.png)

billhollings commented on MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.h:
```diff
@@ -384,17 +385,18 @@ class MVKResourcesCommandEncoderState : public MVKCommandEncoderState {
+        T db = b;   // Copy that can be marked dirty
+        db.markDirty();
         bindings.push_back(db);
     }
```
Minor nit. Could also skip need for `db` copy with:

```
bindings.push_back(b);
bindings.back().markDirty();
```

billhollings commented on MoltenVK/MoltenVK/Commands/MVKMTLResourceBindings.h:
```diff
@@ -49,8 +68,27 @@ typedef struct MVKMTLBufferBinding {
+            isDirty = true;
+        } else if (offset != other.offset) {
+            offset = other.offset;
+            justOffset = !isDirty || justOffset;
```
TBH...I'm having trouble following the logic of when `justOffset` will be set to `true`. Here we seem to be setting it to `true` only if I am not already dirty or `justOffset` was previously set to `true` somehow. 

I may just be missing something in the flow, but it's not clear, in the same way that the values of `isDirty` and `justOffset` are clearly set on each invocation everywhere else.

js6i replied:
My thinking is that if the binding was updated previously, we can't set it to justOffset here, unless the previous update also only changed the offset. If we did, we wouldn't set the changed buffer from the first update when encoding state.

billhollings reviewed (changes requested):
Couple of new questions and comments.

billhollings commented:
Thanks to everyone for proactively testing this PR!

We'll need some form of verification about whether the possible regression is avoided, before pulling it in.

billhollings commented on MoltenVK/MoltenVK/Commands/MVKCommandEncoderState.h:
```diff
@@ -384,17 +385,18 @@ class MVKResourcesCommandEncoderState : public MVKCommandEncoderState {
-                *iter = db;
-                return;
-            }
+            if (iter->index != b.index) continue;
```
Why is this flow not the more direct positive form:

```
if (iter->index == b.index) {
    iter->update(b);
    return;
}
```

js6i replied:
I guess I slightly more like it my way, but I'll change to this next time I update the branch.

Gcenx commented:
@js6i Fleet isn’t using the SE client but the third-party client [XIV on Mac](https://www.xivmac.com/) that’s the same client that Bill used to fix #1592

That uses MoltenVK-CX21-v1.1.3 by default or MoltenVK-v1.1.8 when the “modern” checkbox is checked, it bundles [dxvk-macOS-async-1.10.1](https://github.com/Gcenx/DXVK-macOS/releases/download/v1.10.1/dxvk-macOS-async-1.10.1.tar.gz)

The flickering lighting is best seen in Crystarium  (dosal gate)

My DXVK-macOS packages work with upstream wine64 with stock MoltenVK no problem just use the env options I’d provided in my prior comment.

js6i commented:
Okay, I reproduced the issue with flickering shadows. Presumably the lights was the same problem. The depth-stencil state wasn't getting marked dirty in its beginMetalRenderPass override in the cases handled by the parent method. That should be fixed now.

italomandara commented:
Tested on CX21 with the latest dxvk from gcenx and the mentioned flags turned on.
Titanfall2: no glitches, fps bump from 80-90 to 110-120
Rocket league: no glitches (but the old missing labels bug) fps bump from 90 to 120 fps
Elite dangerous odyssey: no glitches, shader pre-compilation is faster(?) no noticeable in-game fps boost
Guilty gear Xrd: no glitches,, no noticeable fps boost.
looks good so far.
Maybe this is not that much but i wanted to help somehow.

Gcenx commented:
@italomandara the Rocket League nameplate issue is lightly due to the current lack of transform-feedback/geometry shaders support.

I’d asked @seathasky to test this and they’d confirmed the shadows regression was indeed resolved, however the lighting bug still remains that’s been the case for any version except MoltenVK shipped in CX20.

Gcenx commented:
@js6i not sure if you’ve seen my comment.

The shadows regression was resolved, the lighting wasn’t resolved. From my perspective it’s not worth holding their back from getting merged just for the FFXIV lightly problem.

js6i commented:
> @js6i not sure if you’ve seen my comment.
> 
> The shadows regression was resolved, the lighting wasn’t resolved. From my perspective it’s not worth holding their back from getting merged just for the FFXIV lightly problem.

Wait, I thought you meant that the lighting problem was introduced earlier. If that's not the case, and it is in fact caused by changes in this PR, it should definitely be fixed before merging. In such case I'd ask for directions to this Crystorium place, or somewhere else where I'm going to see it easily.

Gcenx commented:
@js6i this PR didn’t cause the lighting problem that was prior behavior, the shadows problem was caused by and fix in this PR.

The lighting problem has been around since 1.1.2, as 1.1.1 was the last version without that problem see #1542

FleetAdmiralButter commented:
Hey @js6i,

Thanks for looking into this! I grabbed the latest build from Actions but it's unfortunately still exhibiting the same behaviour:

![image](https://user-images.githubusercontent.com/37036084/170808384-8e3b6ce3-261a-4e3b-90b4-f129b0827188.png)

FleetAdmiralButter commented:
Gave this a try on my M1 Pro and got a similar result:

<img width="1559" alt="image" src="https://user-images.githubusercontent.com/37036084/170815948-bdda91b6-1220-4fd8-998a-d680e90fbdef.png">

seathasky commented:
Tested with RX 5700 XT on MacOS Big Sur gives this result: 
![image](https://user-images.githubusercontent.com/17598024/170817751-4bba207a-9e9d-46ab-80f7-e7b8fa960133.png)

seathasky commented:
Suprising this issue doesnt affect Polaris based card (RX500 Series), but does affect M1 and Navi based AMD card. 

One thing to note, on Polaris (500 series), the regressions and stuttering are still very apparent and unplayable with this build.

seathasky commented:
Here is video footage of 5700 XT on Big Sur:

https://user-images.githubusercontent.com/17598024/170817906-0f5475b1-05d1-4088-8189-faf33b60cb5b.mp4

seathasky commented:
Here is footage of **(Polaris) RX570** with same build:

https://user-images.githubusercontent.com/17598024/170818080-853855a6-7577-4b28-bd1f-d1461227bb70.mp4

seathasky commented:
> Here is footage of **(Polaris) RX570** with same build:
> 2022-05-28-03-54-31.1.mp4
> 
> Stutter regression remains there but lighting seems fine compared to Navi.

I need to point something out very important here that I forgot to mention in my initial post about AMD 500 series. 

@billhollings There has yet to be a release of MoltenVK from your page that has worked properly for 500 series. 

Every single release has introduced stuttering to these 500 series cards. Jeremy over at Codeweavers provided the us, the "XIV on Mac" launcher team distribution of one of there CX 21 MVK dylibs that solved this issue entirely. 

It's actually the release we currently use as our "stable" release that ships with the XIV on Mac launcher. I just didnt want to confuse anyone.

Gcenx commented:
@billhollings I'm bugging them on discord now

Gcenx commented:
> Are there lighting/shadow issues that are being introduced with this PR, or is it safe to pull in, relative to existing MoltenVK behavior?

The only thing to look into having in upstream is explained by Chip https://github.com/KhronosGroup/MoltenVK/issues/1597

seathasky commented:
We can ignore the 500 series issue for now. One thing I will say about it though is Codeweavers did provide the source. We weren't able to achieve the same results as their compiled dylib after compiling their source. Maybe they sent the wrong source? Something to do with xcode? Who knows.

**On to this PR.** 

I've done some extensive testing in FF14 today with my 5700xt and the latest pr. Hours of testing actually. The PR is extremely stable, but it still has flickering in certain places. It's interesting too. In the Dossal gate, you can see it strobes faster than usual. 


Short video:
https://www.youtube.com/watch?v=IpD-y6Ouyks

js6i commented:
@billhollings I don't reproduce any more issues with AMD Radeon Pro 5500M. I want to test it on M1 too, given the possibility of something hardware dependent, but I couldn't get it to run right away and was busy with some other work. Let me ping you once I manage that.

FleetAdmiralButter commented:
Can confirm the latest commit fixes the lighting issues for me as well, thanks @js6il.

js6i commented:
> Can confirm the latest commit fixes the lighting issues for me as well, thanks @js6il.

That's confusing, because I didn't change anything since your last report!

Anyways, @billhollings, this time running the game on M1 worked and I didn't see any lighting issues there either. I found a bug in `MVKDepthBiasCommandEncoderState::setDepthBias` that is fixed in the latest version, however that wasn't causing any visible trouble. If anyone's still seeing regressions caused by these changes, let me know, but I couldn't reproduce any - sorry!

Gcenx commented:
@js6i your changed had introduced a regression with shadows but you’ve already resolved that.

The other thing mentioned here was #1542 that’s a rather long standing bug. Your changes have no made any difference to that issue.

<br>

I hope this helps clear things up for you.

billhollings commented:
@cdavis5e It's been a bit since your review. All good now from your perspective?
